storage-resize-images/functions/src/index.ts

```diff
@@ -49,149 +47,194 @@ logs.init();
+      if (remoteFile) {
+        try {
+          logs.remoteFileDeleting(filePath);
+          await remoteFile.delete();
+          logs.remoteFileDeleted(filePath);
+        } catch (err) {
+          logs.errorDeleting(err);
+        }
+      }
     }
+  }
+};
 
-    if (
-      config.includePathList &&
-      !startsWithArray(config.includePathList, tmpFilePath)
-    ) {
-      logs.imageOutsideOfPaths(config.includePathList, tmpFilePath);
-      return;
-    }
+export const generateResizedImage = functions.storage
+  .object()
+  .onFinalize(async (object) => {
+    await generateResizedImageHandler(object);
+  });
 
-    if (
-      config.excludePathList &&
-      startsWithArray(config.excludePathList, tmpFilePath)
-    ) {
-      logs.imageInsideOfExcludedPaths(config.excludePathList, tmpFilePath);
+/**
+ *
+ */
+export const backfillResizedImages = functions.tasks
+  .taskQueue()
+  .onDispatch(async (data) => {
+    const runtime = getExtensions().runtime();
+    if (!config.doBackfill) {
+      await runtime.setProcessingState(
+        "PROCESSING_COMPLETE",
+        "Existing images were not resized."
+      );
       return;
     }
-
-    if (object.metadata && object.metadata.resizedImage === "true") {
-      logs.imageAlreadyResized();
-      return;
+    if (data == undefined) {
+      logs.startBackfill();
     }
-
-    const bucket = admin.storage().bucket(object.bucket);
-    const filePath = object.name; // File path in the bucket.
-    const parsedPath = path.parse(filePath);
-    const objectMetadata = object;
-
-    let originalFile;
-    let remoteFile;
-    try {
-      originalFile = path.join(os.tmpdir(), filePath);
-      const tempLocalDir = path.dirname(originalFile);
-
-      // Create the temp directory where the storage file will be downloaded.
-      logs.tempDirectoryCreating(tempLocalDir);
-      await mkdirp(tempLocalDir);
-      logs.tempDirectoryCreated(tempLocalDir);
-
-      // Download file from bucket.
-      remoteFile = bucket.file(filePath);
-      logs.imageDownloading(filePath);
-      await remoteFile.download({ destination: originalFile });
-      logs.imageDownloaded(filePath, originalFile);
-
-      // Get a unique list of image types
-      const imageTypes = new Set(config.imageTypes);
-
-      // Convert to a set to remove any duplicate sizes
-      const imageSizes = new Set(config.imageSizes);
-
-      const tasks: Promise<ResizedImageResult>[] = [];
-
-      imageTypes.forEach((format) => {
-        imageSizes.forEach((size) => {
-          tasks.push(
-            modifyImage({
-              bucket,
-              originalFile,
-              parsedPath,
-              contentType,
-              size,
-              objectMetadata: objectMetadata,
-              format,
-            })
-          );
-        });
+    const bucket = admin.storage().bucket(process.env.IMG_BUCKET);
+    const query = data.nextPageQuery || {
+      autoPaginate: false,
+      maxResults: 3,
```

Comment:
Why maximum of 3? That seems low? We should do as many as we can per chunk of work while still being guaranteed to finish resizing them in 9 minutes. The free tier for task executions is generous, but not infinite :)

Reply:
This is to avoid out of memory errors when images are larger. Even when resizing single images, we've gotten reports of OOM errors (hence the param to control memory).

Also, when thinking about price concerns, i think its important to consider what the biggest costs will be. In this case, each task/function execution is going to do a lot of data egress/ingress in and out of Cloud Storage, and will consume a lot of compute time. If a user is exceeding the task execution free tier (1m executions) for even very small images (ie 1MB per image), they would incur the following charges:
- 3 million object inserts (Class A operation, $.05 per 10000) = $15
- 3 million object gets (Class B operation, $.004 per 10000) = $1.2
- 3 million MB storage (cost depends on region, but cheapest is ~$.02 per gb per month) = $60 per month
- as well as charge for the compute time of the functions

Tasks only cost $.4 per million after the first million.  With our current set up, this extension will incur at very least $75 of other costs for every $.4 spent on tasks.